### PR TITLE
[WIP] Adds the CEK machine to plutus-core-interpreter

### PR DESCRIPTION
**Pull request by effectfully**

### CODE REVIEW

**Comment by michaelpj** [comment]
(Let me know when this is ready for review)

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+computeCek env con (TyInst _ fun ty)      = computeCek env (FrameTyInstArg ty : con) fun
+computeCek env con (Apply _ fun arg)      = computeCek env (FrameApplyArg env arg : con) fun
+computeCek env con (Wrap ann tyn ty term) = computeCek env (FrameWrap ann tyn ty : con) term
+computeCek env con (Unwrap _ term)        = computeCek env (FrameUnwrap : con) term
+computeCek env con tyAbs@TyAbs{}          = returnCek env con tyAbs
+computeCek env con lamAbs@LamAbs{}        = returnCek env con lamAbs
+computeCek env con constant@Constant{}    = returnCek env con constant
+computeCek _   _   Error{}                = EvaluationFailure
+computeCek env con var@(Var _ name)       = case lookupName name env of
+    Nothing           -> throw $ MachineException OpenTermEvaluatedMachineError var
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: To elaborate on this: imagine someone using this machine to evaluate Plutus Core given to them by users. They do no want to catch exceptions, ideally they're crashing conditions indicating bugs. But the user might give them an open term, so they'd like to be robust to that and have it be an error they can handle normally.

**Comment by effectfully** [comment]
@michaelpj, requested a review. Do not look at commits, just look at "files changes". If you're going to merge this, squash everything.

**Comment by michaelpj** [comment]
Will do. You should feel free to rebase and force-push in this situation - it will do the right thing.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+import           Data.IntMap                                     (IntMap)
+import qualified Data.IntMap                                     as IntMap
+
+type Plain f = f TyName Name ()
+
+-- | Environments used by the CEK machine.
+-- Each row is a mapping from the 'Unique' representing a variable to the 'Value' the
+-- variable stands for and the environment the value was defined in.
+newtype Environment = Environment (IntMap (Plain Value, Environment))
```
Comment (michaelpj) [inline plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I'd be tempted to do `type Closure = (Plain Value, Environment)`.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+computeCek env con (Unwrap _ term)        = computeCek env (FrameUnwrap : con) term
+computeCek env con tyAbs@TyAbs{}          = returnCek env con tyAbs
+computeCek env con lamAbs@LamAbs{}        = returnCek env con lamAbs
+computeCek env con constant@Constant{}    = returnCek env con constant
+computeCek _   _   Error{}                = EvaluationFailure
+computeCek env con var@(Var _ name)       = case lookupName name env of
+    Nothing           -> throw $ MachineException OpenTermEvaluatedMachineError var
+    Just (term, env') -> returnCek env' con term
+
+-- | The returning part of the CEK machine.
```
Comment (michaelpj) [inline plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I found it helpful to realise that this function always pops a frame off the stack (unless it's done).

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+-- If the function is not a 'LamAbs', then 'Apply' it to the argument and view this
+-- as an iterated application of a 'BuiltinName' to a list of 'Value's.
+-- If succesful, proceed with either this same term or with the result of the computation
+-- depending on whether 'BuiltinName' is saturated or not.
+applyEvaluate :: Environment -> Environment -> Context -> Plain Value -> Plain Value -> EvaluationResult
+applyEvaluate funEnv argEnv con (LamAbs _ name _ body) arg =
+    computeCek (extendEnvironment name arg argEnv funEnv) con body
+applyEvaluate funEnv _      con fun                    arg =
+    let term = Apply () fun arg in
+        case termAsPrimIterApp term of
```
Comment (michaelpj) [inline plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: again, some judicious `ViewPatterns` might make this nicer.

File: plutus-core-interpreter/test/CekMachine.hs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import           Hedgehog                                   hiding (Size, Var)
+import           Test.Tasty
+import           Test.Tasty.Hedgehog
+
+-- TODO: also type check the terms.
+propTermOfTypedBuiltinValue :: GenT Quote (TermOf (TypedBuiltinValue Size a)) -> Property
+propTermOfTypedBuiltinValue genTermOfTbv = property . hoist (return . runQuote) $ do
+    TermOf term tbv <- forAllPrettyCfgT genTermOfTbv
+    resExpected <- lift $ unsafeMakeBuiltin tbv
+    for_ (evaluationResultToMaybe $ evaluateCek term) $ \resActual ->
```
Comment (michaelpj) [inline plutus-core-interpreter/test/CekMachine.hs]: I'm nitpicking now, but this function is also duplicated and just takes a generic evaluator. Perhaps it could go in the support module?

File: stack.yaml [hunk 6]
```diff
@@ -3,6 +3,7 @@ resolver: lts-11.11
 packages:
 - plutus-prototype
 - language-plutus-core
+- plutus-core-interpreter
```
Comment (michaelpj) [inline stack.yaml]: also needs to go in default.nix and update the pkgs

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+extendEnvironment :: Name () -> Plain Value -> Environment -> Environment -> Environment
+extendEnvironment argName arg argEnv (Environment oldEnv) =
+    Environment $ IntMap.insert (unUnique $ nameUnique argName) (arg, argEnv) oldEnv
+
+-- | Look up a name in an environment.
+lookupName :: Name () -> Environment -> Maybe (Plain Value, Environment)
+lookupName name (Environment env) = IntMap.lookup (unUnique $ nameUnique name) env
+
+-- | The computing part of the CEK machine.
+computeCek :: Environment -> Context -> Plain Term -> EvaluationResult
```
Comment (michaelpj) [inline plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: No operators this time? ;)

In all seriousness, I think it would be nice if this was as close to the CEK machine as possible, stylistically too. E.g. they both have a computing and a returning part, and a context, etc.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Evaluation.MachineException (MachineError (..), MachineException (..))
+import           Language.PlutusCore.Evaluation.Result           (EvaluationResult (..))
+import           Language.PlutusCore.View
+import           PlutusPrelude
+
+import           Data.IntMap                                     (IntMap)
+import qualified Data.IntMap                                     as IntMap
+
+type Plain f = f TyName Name ()
```
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I'm not sure about what to do here. I do not want to pollute the global namespace with this non-descriptive name. Do you have any name suggestions?
We can turn it into a type family and make it work over `Type` too :)

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I think we should keep the CK machine as is. It's a reference interpreter and must be as simple as possible. Returning explicit errors would complicate it to a certain extent. But you're right, the CEK machine should handle this situation gracefully.
Then we also need to split exceptions into user and internal ones.

[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: Although... an open term won't type check, right?

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+returnCek :: Environment -> Context -> Plain Value -> EvaluationResult
+returnCek _   []                             res = EvaluationSuccess res
+returnCek env (FrameTyInstArg ty      : con) fun = instantiateEvaluate env con ty fun
+returnCek env (FrameApplyArg env' arg : con) fun = computeCek env' (FrameApplyFun env fun : con) arg
+returnCek env (FrameApplyFun env' fun : con) arg = applyEvaluate env' env con fun arg
+returnCek env (FrameWrap ann tyn ty   : con) val = returnCek env con $ Wrap ann tyn ty val
+returnCek env (FrameUnwrap            : con) dat = case dat of
+    Wrap _ _ _ term -> returnCek env con term
+    term            -> throw $ MachineException NonWrapUnwrappedMachineError term
```
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: Should be impossible, yes.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+-- | Instantiate a term with a type and proceed.
+-- In case of 'TyAbs' just ignore the type. Otherwise check if the term is an
+-- iterated application of a 'BuiltinName' to a list of 'Value's and, if succesful,
+-- apply the term to the type via 'TyInst'.
+instantiateEvaluate
+    :: Environment -> Context -> Type TyName () -> Plain Term -> EvaluationResult
+instantiateEvaluate env con _  (TyAbs _ _ _ body) = computeCek env con body
+instantiateEvaluate env con ty fun
+    | isJust $ termAsPrimIterApp fun = returnCek env con $ TyInst () fun ty
```
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I usually do not use `ViewPatterns`, because they're kinda sneaky and I prefer to be explicit about where I compute. It's fine for me either way, but I like the way it's currently written more.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    | otherwise                      = throw $ MachineException NonPrimitiveInstantiationMachineError fun
```
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: Yes, not supposed to happen.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 2 above)
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I think I'm going to make this an actual data type then.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: Or we could be very boring and have two type synonyms ;)
```
type PlainTerm = Term TyName Name ()
type PlainType = Type TyName ()
```

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 4 above)
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: Could you show? It's maybe slightly verbose currently, but is kinda explicit about what is happening. Like "received a function, applied it to an argument, got a term, viewed it as a primitive iterated application". Easy to read.

File: plutus-core-interpreter/test/CekMachine.hs (hunk 5 above)
Comment (effectfully) [reply plutus-core-interpreter/test/CekMachine.hs]: Agreed.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I'm not sure if an open term will typecheck currently, but we might want to allow it to. 

And maybe we won't always typecheck what the user gives us? I'd still be inclined to err on the side of returning errors due to malformed input too - they're "user errors" by the user of the function, and I'd still prefer to only throw if it's *our* error.

Re: user errors and internal ones, we can make the user error type an instance of `Exception` and then throw it in the CK machine and return it in the `CEK` machine :)

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I like it in this kind of case because the check is conceptually "part of" the match, but I don't mind if you don't like them.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 2 above)
Comment (michaelpj) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: :+1:

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 7 above)
Comment (michaelpj) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: er, the CK machine

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: Also

```haskell
type PlainValue = Value TyName Name ()
type PlainProgram = Program TyName Name ()
```

I'm fine with this.

[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: But that changes a lot across the code base, so belongs to another PR.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+-- except we do not use substitution and use environments instead.
+-- The CEK machine relies on variables having
+-- 1. equal 'Unique's whenever they have equal string names
+-- 2. non-equal 'Unique's whenever they have non-equal string names
+-- I.e. 'Unique's are used instead of string names, so a renamer pass is required.
+-- This is for efficiency reasons.
+-- The CEK machine generates booleans along the way which might contain globally non-unique 'Unique's.
+-- This is not a problem as the CEK machines handles name capture by design.
+
+module PlutusCoreInterpreter.CekMachine
```
Comment (michaelpj) [inline plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: Oh, and I don't like this module prefix. At least `PlutusCore.Interpreter`. Ideally with the `Language` prefix we're using elsewhere.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 4 above)
Comment (michaelpj) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: Ah, I take it back. You're using two variables from the original pattern, so it's not so simple.

This is the best I've got, not very compelling:
```
diff --git i/plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs w/plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs
index 6835aa5..4b96b94 100644
--- i/plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs
+++ w/plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs
@@ -9,6 +9,8 @@
 -- The CEK machine generates booleans along the way which might contain globally non-unique 'Unique's.
 -- This is not a problem as the CEK machines handles name capture by design.
 
+{-# LANGUAGE ViewPatterns #-}
+
 module PlutusCoreInterpreter.CekMachine
     ( EvaluationResult (..)
     , evaluateCek
@@ -96,17 +98,15 @@ instantiateEvaluate env con ty fun
 applyEvaluate :: Environment -> Environment -> Context -> Plain Value -> Plain Value -> EvaluationResult
 applyEvaluate funEnv argEnv con (LamAbs _ name _ body) arg =
     computeCek (extendEnvironment name arg argEnv funEnv) con body
-applyEvaluate funEnv _      con fun                    arg =
-    let term = Apply () fun arg in
-        case termAsPrimIterApp term of
-            Nothing                       ->
-                throw $ MachineException NonPrimitiveApplicationMachineError term
-            Just (IterApp headName spine) ->
-                case runQuote $ applyBuiltinName headName spine of
-                    ConstAppSuccess term' -> returnCek funEnv con term'
-                    ConstAppFailure       -> EvaluationFailure
-                    ConstAppStuck         -> returnCek funEnv con term
-                    ConstAppError err     -> throw $ MachineException (ConstAppMachineError err) term
+applyEvaluate funEnv _      con fun                    arg = let term = Apply () fun arg in
+    case term of
+        (termAsPrimIterApp -> Just (IterApp headName spine)) ->
+            case runQuote $ applyBuiltinName headName spine of
+                ConstAppSuccess term' -> returnCek funEnv con term'
+                ConstAppFailure       -> EvaluationFailure
+                ConstAppStuck         -> returnCek funEnv con term
+                ConstAppError err     -> throw $ MachineException (ConstAppMachineError err) term
+        _ -> throw $ MachineException NonPrimitiveApplicationMachineError term
 
 -- | Evaluate a term using the CEK machine. May throw a 'MachineException'.
 evaluateCek :: Term TyName Name () -> EvaluationResult
```

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I don't think we should enforce using them , we can just put them there if you want them. Many places can't use them because they're generic over the annotation.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: We will always rename, though, because the CEK machine relies on a renamer pass. And we have there

```haskell
-- | A 'RenameError' is thrown when a free variable is encountered during
-- rewriting.
data RenameError a = UnboundVar (Name a)
                   | UnboundTyVar (TyName a)
                   deriving (Show, Eq, Generic, NFData)
```

So it's our error if there is a free variable and the renamer hasn't caught it.

[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: Okay, fair enough. Let's just make sure we don't expose an interface to a library user where they can get this wrong, then.

File: stack.yaml (hunk 6 above)
Comment (effectfully) [reply stack.yaml]: What should I do with `default.nix`?

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I tried and disliked what I saw.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 4 above)
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: Yeah, not too nice.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: What if someone wants to run a computation twice, but rename only once? Like when we put something on the blockchain (or is this not the case?).

File: stack.yaml (hunk 6 above)
Comment (effectfully) [reply stack.yaml]: I'm not sure. `plutus-core-interpreter` is of the same kind as `language-plutus-core`, but the latter only has a line about benchmarks. And I'm getting `[2 built, 0.0 MiB DL]` after I run `nix build -f default.nix plutus-core-interpreter`.

Comment (effectfully) [reply stack.yaml]: Oh, there is another `default.nix`, namely the one in `pkgs`, but I updated it using `pkgs/generate.sh`.

Comment (michaelpj) [reply stack.yaml]: You can think of those finding functions as `Package -> Package`. You don't want the benchmark call because you don't have any, hence why I suggested copying `plutus-prototype`.

You did replace the second occurrence of `plutus-prototype` in the kin line, right?

Comment (michaelpj) [reply stack.yaml]: You can also run `nix log` with the same arguments to see the build log and check it's doing what you expect.

Comment (michaelpj) [reply stack.yaml]: No, I mean the one in the root, the other one is indeed generated.

Comment (effectfully) [reply stack.yaml]: `nix log` gives me a huge wall of text, but I see stuff like:

```
Building library for plutus-core-interpreter-0.1.0.0..
[1 of 1] Compiling Language.PlutusCore.Interpreter.CekMachine ( src/Language/PlutusCore/Interpreter/
ignoring (possibly broken) abi-depends field for packages
Preprocessing test suite 'plutus-core-interpreter-test' for plutus-core-interpreter-0.1.0.0..
Building test suite 'plutus-core-interpreter-test' for plutus-core-interpreter-0.1.0.0..
[1 of 2] Compiling CekMachine       ( test/CekMachine.hs, dist/build/plutus-core-interpreter-test/pl
[2 of 2] Compiling Main             ( test/Main.hs, dist/build/plutus-core-interpreter-test/plutus-c
Linking dist/build/plutus-core-interpreter-test/plutus-core-interpreter-test ...
@nix { "action": "setPhase", "phase": "checkPhase" }
running tests
Running 1 test suites...
Test suite plutus-core-interpreter-test: RUNNING...
Test suite plutus-core-interpreter-test: PASS
```

I didn't touch the `default.nix` file. Also as I said above there is only one line related to `language-plutus-core` and it's about benchmarks, so maybe `default.nix` just shouldn't be touched, because `plutus-core-interpreter` is similar to `language-plutus-core` modulo benchmarks.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I thought about this a bit more and I think it doesn't make sense to return explicit errors in case an ill-typed terms was fed to the CK machine. Because
1) if you return an `Either`, it means the user must handle the `Left` case, i.e. handle the possibility of causing an internal error which is silly
2) the user is essentially forced to recheck that a term indeed was type checked before evaluated
3) the user can simply ignore the `Left` case and we won't know something goes wrong from time to time

There is nothing wrong about having invariants and punishing by throwing exceptions those who break them.

File: stack.yaml (hunk 6 above)
Comment (michaelpj) [reply stack.yaml]: Here's how it goes:
- Adding it to `pkgs/default.nix` via `generate.sh` puts it in the package set (sorry, I was wrong: I implied that you needed to edit the root `default.nix` before this would happen!)
- The overlay in `/default.nix` tweaks a few properties of our packages.
    - `plutus-prototype` has `addRealTimeLogs` and `filterSource` added.
    - `language-plutus-core` additionally has some stuff relating to benchmarks turned on, and haddocks.
    - We definitely want the first two for `plutus-core-interpreter`, but you probably also want the haddocks.

Here's a diff to apply:
```
diff --git i/default.nix w/default.nix
index 3cf2902..b827893 100644
--- i/default.nix
+++ w/default.nix
@@ -54,6 +54,7 @@ let
       plutus-prototype = addRealTimeTestLogs (filterSource super.plutus-prototype);
       # we want to enable benchmarking, which also means we have criterion in the corresponding env
       language-plutus-core = appendConfigureFlag (doBenchmark (doHaddockHydra (addRealTimeTestLogs (filterSource super.language-plutus-core)))) "--enable-benchmarks";
+      plutus-core-interpreter = doHaddockHydra (addRealTimeTestLogs (filterSource super.plutus-core-interpreter));
     };
   };
   other = rec {
```

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: I think your example of nodes maybe wanting to evaluate without typechecking is a good one. More generally, doing it that way gives people the choice. Given a CEK machine that returns an `Either`, we can then implement a version that doesn't return an either and instead throws, possibly requiring a typechecked term. And then users who don't want to do that naturally have to use the version that returns `Either`.

I just don't like locking the error handling strategy to "exception" unless they literally can only happen if they're a bug, which is not currently the case with the machine failures.

File: plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs [hunk 13]
```diff
@@ -0,0 +1,118 @@
+-- | The CEK machine.
+-- Rules are the same as for the CK machine from "Language.PlutusCore.Evaluation.CkMachine",
+-- except we do not use substitution and use environments instead.
+-- The CEK machine relies on variables having
+-- 1. equal 'Unique's whenever they have equal string names
```
Comment (effectfully) [inline plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs]: This is wrong.

File: stack.yaml (hunk 6 above)
Comment (effectfully) [reply stack.yaml]: Thank you, I understand now. Fixed.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: > I just don't like locking the error handling strategy to "exception" unless they literally can only happen if they're a bug, which is not currently the case with the machine failures.

A renamer pass is a must, because otherwise you can get an `EvaluationSuccess` with a wrong result. Insisting on type checking terms is no harder than insisting on renaming them. I think making the type checker return a `WtTerm` turns machine failures into true bugs and resolves the issue.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 7 above)
Comment (michaelpj) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: My preference would then be to use `computeCk` and `returnCk` in the CK machine, but not a big deal.

File: default.nix [hunk 14]
```diff
@@ -54,6 +54,7 @@ let
       plutus-prototype = addRealTimeTestLogs (filterSource super.plutus-prototype);
       # we want to enable benchmarking, which also means we have criterion in the corresponding env
       language-plutus-core = appendConfigureFlag (doBenchmark (doHaddockHydra (addRealTimeTestLogs (filterSource super.language-plutus-core)))) "--enable-benchmarks";
+	  plutus-core-interpreter = doHaddockHydra (addRealTimeTestLogs (filterSource super.plutus-core-interpreter));
```
Comment (michaelpj) [inline default.nix]: indentation?

Comment (effectfully) [reply default.nix]: An accidental tab.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 7 above)
Comment (effectfully) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: It's more important for a reference interpreter to match the specification it's from rather than subsequent interpreters.

**Comment by effectfully** [comment]
Ok, it seems I've fixed everything except for the explicit errors vs exceptions issue. The gist of what I think:

> > I just don't like locking the error handling strategy to "exception" unless they literally can only happen if they're a bug, which is not currently the case with the machine failures.

> A renamer pass is a must, because otherwise you can get an `EvaluationSuccess` with a wrong result. Insisting on type checking terms is no harder than insisting on renaming them. I think making the type checker return a `WtTerm` turns machine failures into true bugs and resolves the issue.

@michaelpj, if you disagree with this position, we can ask the others what they think.

File: plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs (hunk 7 above)
Comment (mchakravarty) [reply plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs]: In general I agree with the sentiment to try to make the two as similar as possible, but in this case, I think, the overriding priorities ought to be that (1) the CK machine is as close to the LaTeX in the spec as possible (as @effectfully says) and (2) the CEK machine being good production Haskell code. We'll test CEK against CK; so, that is how we'll get them to match up.

**Comment by mchakravarty** [comment]
@effectfully wrote,
> Ok, it seems I've fixed everything except for the explicit errors vs exceptions issue. The gist of what I think:
> 
> > > I just don't like locking the error handling strategy to "exception" unless they literally can only happen if they're a bug, which is not currently the case with the machine failures.
> 
> > A renamer pass is a must, because otherwise you can get an `EvaluationSuccess` with a wrong result. Insisting on type checking terms is no harder than insisting on renaming them. I think making the type checker return a `WtTerm` turns machine failures into true bugs and resolves the issue.

To be honest, I don't understand this. I agree with @michaelpj that we only want exceptions for fatal, unrecoverable (usually internal) errors. Everything else ought to be an error value. (It was me suggesting this distinction in the first place, and we probably want to add it to the contribution guidelines.) 

I am happy to make an exception for the CK machine to keep it's code close to the spec, but I don't see any reason for that for the CEK machine.

What does this have to do with the rename?

**Comment by michaelpj** [comment]
So I think our disagreement is over what counts as a "fatal, unrecoverable (usually internal)" error. For example, if one feeds a non-typechecked program to the CEK machine, you might hit an error that is currently signalled by an exception.

My interpretation is that this is a user error (where the user is the caller of the function), and so should not be an exception.

(Personally I would go further and say exceptions should only be used for "impossible" situations that indicate a bug in the code that we wrote, but maybe that's me.)

**Comment by effectfully** [comment]
> I agree with @michaelpj that we only want exceptions for fatal, unrecoverable (usually internal) errors. Everything else ought to be an error value.

I agree with this attitude.

> What does this have to do with the renamer?

We cannot catch bugs provided there was no renamer pass. No renamer pass = code can potentially return wrong results without throwing/returning errors. Thus the user is required to perform renaming and we cannot control that (except by newtypes, but that's irrelevant for now). Once we introduce this invariant, we can as well just say "type check your code before feeding it to the CEK machine". That's a single invariant just like the renaming one, but it also subsumes the latter. Provided the code was type checked, the CEK machine can only fail in the case of an internal error. Evaluation of a well-typed term cannot result in an error unless the error is internal. So if we have no choice, but to enforce some invariant, why not enforce the invariant that frees the user from handling errors? It is simply safer, because with such an API the user is not allowed to mess with internal exceptions and swallow them for example.

To summarize, we have two options:
1. enforce the invariant that the renamer pass must precede evaluation. Return an `Either` where the `Left` case is due to either internal error or the user breaking the type checking invariant
2. enforce the invariant that the type checker pass must precede evaluation. Return pure `EvaluationResult`, because any error occurring during evaluation is an internal one

To me (2) is the clear winner.

**Comment by michaelpj** [comment]
What about
3. Expose a variant that returns `Either`, and a variant that takes a typechecked term and throws any errors as exceptions.

**Comment by effectfully** [comment]
> 3. Expose a variant that returns `Either`, and a variant that takes a typechecked term and throws any errors as exceptions.

Do you propose to provide both the (1) and (2) APIs? I do not see a reason for that as (2) is the clear winner and should be the only API.

**Comment by michaelpj** [comment]
I don't think that's obvious. I still think 1. is better. In fact I think we should return errors even for the things like open terms.

**Comment by michaelpj** [comment]
I just don't see any reason not to do it - everyone doing this is going to be in a context where they're handling errors anyway.

**Comment by effectfully** [comment]
> In fact I think we should return errors even for the things like open terms.

We do return them in the renamer. 

> I just don't see any reason not to do it - everyone doing this is going to be in a context where they're handling errors anyway.

Because

1. once you managed to enforce the right invariant, all errors become internal and you can save the user from ever seeing them unless such an error occurs in runtime
2. you're forced to enforce some invariant in any case

So it's all about which invariant leads to better API. The API of (2) is more convenient than the API of (1) and hence (2) wins.

**Comment by effectfully** [comment]
@mchakravarty, I explained my position, so please just choose a solution and I'll implement it whether I like it or not.

**Comment by michaelpj** [comment]
I don't think it's a big enough deal to hold this up - if we do settle on a firm policy it's very straightforward to revisit this.
